src/pexprs-eval.js

```diff
@@ -285,9 +280,11 @@ pexprs.Apply.prototype.eval = function(state) {
 
   var memoKey = app.toMemoKey();
   var memoRec = posInfo.memo[memoKey];
+  var isTopLevelApplication =
+      !caller && this !== state.applySpaces && this !== pexprs.end;
```

Comment:
Nit: `pexprs.end` is not really an application, so it seems weird to have it show up in this expression. How about

    var isTopLevelApplication = !caller && this instanceof pexprs.Apply && this !== state.applySpaces;

instead?